mirgecom/boundary.py

```diff
@@ -49,86 +49,63 @@
 from meshmode.mesh import BTAG_ALL, BTAG_NONE  # noqa
 from grudge.dof_desc import as_dofdesc
 from grudge.symbolic.primitives import TracePair
-from mirgecom.fluid import split_conserved, join_conserved
+from mirgecom.fluid import (
+    split_conserved,
+    join_conserved
+)
 from mirgecom.inviscid import inviscid_facial_flux
+from mirgecom.viscous import viscous_facial_flux
+from mirgecom.artificial_viscosity import AVBoundaryInterface
+from mirgecom.navierstokes import NSBoundaryInterface
 
 
-from abc import ABCMeta, abstractmethod
+class AVNSBoundaryInterface(NSBoundaryInterface, AVBoundaryInterface):
+    r"""Base interface to Navier-Stokes+ArtVisc boundary treatments."""
 
+    pass
 
-class FluidBoundary(metaclass=ABCMeta):
-    r"""Abstract interface to fluid boundary treatment.
-
-    .. automethod:: inviscid_boundary_flux
-    .. automethod:: viscous_boundary_flux
-    .. automethod:: q_boundary_flux
-    .. automethod:: s_boundary_flux
-    .. automethod:: t_boundary_flux
-    """
-
-    @abstractmethod
-    def inviscid_boundary_flux(self, discr, btag, q, eos, **kwargs):
-        """Get the inviscid flux across the boundary faces."""
-
-    @abstractmethod
-    def viscous_boundary_flux(self, discr, btag, q, eos, **kwargs):
-        """Get the inviscid flux across the boundary faces."""
-
-    @abstractmethod
-    def q_boundary_flux(self, discr, btag, q, eos, **kwargs):
-        """Get the scalar conserved quantity flux across the boundary faces."""
-
-    @abstractmethod
-    def s_boundary_flux(self, discr, btag, q, eos, **kwargs):
-        r"""Get $\nabla\mathbf{Q}$ flux across the boundary faces."""
-
-    @abstractmethod
-    def t_boundary_flux(self, discr, btag, q, eos, **kwargs):
-        r"""Get temperature flux across the boundary faces."""
 
+class FluidBoundary(AVNSBoundaryInterface):
```

Comment:
I prefer this approach to presenting and/or maintaining separate boundary types which implement essentially equivalent boundary treatments depending on use-case.  

Our targeted use-cases have factored in here too.  We will soon almost never use inviscid-only runs.  All of our current BCs will work in both Euler and NS. Euler uses a subset of NS boundaries - and almost a complete set depending on whether AV is used.  All viscous BC's require the inviscid interface (get_inviscid_flux) so all viscous type BCs will work in in the inviscid setting.

IF (ever) someone wants to make an inviscid-only boundary treatment, it is no problem; they just inherit this boundary type and do not implement the viscous interfaces. The boundary device will automatically toss an exception with descriptive message if one tries to use it in a viscous-enabled operator. Or .. just inherit the inviscid interface only, which would also work.

Reply:
> I prefer this approach to presenting and/or maintaining separate boundary types which implement essentially equivalent boundary treatments depending on use-case.

False dichotomy. 🙂 You don't need this base class to make BCs that work with both Euler and NS. You just have them inherit both `EulerBoundaryInterface` and `NSBoundaryInterface` and implement the required operators.

> IF (ever) someone wants to make an inviscid-only boundary treatment, it is no problem; they just inherit this boundary type and do not implement the viscous interfaces. The boundary device will automatically toss an exception with descriptive message if one tries to use it in a viscous-enabled operator.

This is the sign of a code smell IMO. Interfaces shouldn't be used this way.

> Or .. just inherit the inviscid interface only, which would also work.

Yeah. I'm just not (yet) seeing what adding this base class really buys us.

Reply:
> > I prefer this approach to presenting and/or maintaining separate boundary types which implement essentially equivalent boundary treatments depending on use-case.
> 
> False dichotomy. 🙂 You don't need this base class to make BCs that work with both Euler and NS. You just have them inherit both `EulerBoundaryInterface` and `NSBoundaryInterface` and implement the required operators.
> 

Not quite.  The NSBoundaryInterace includes an inviscid interface already, a boundary device should not inherit both.  The NSBoundaryInterace already inherits the InviscidBoundaryInterace. 

The thing that I do not like with (your) suggested approach is that it requires every boundary device to inherit the same N things.   I prefer the approach of proving a base thing that is all N things, then inheriting that base thing in all the boundaries.  As I mentioned, this centralizes the spot where we can come in and add operator support to existing boundaries.

> 
> This is the sign of a code smell IMO. Interfaces shouldn't be used this way.
> 
> Yeah. I'm just not (yet) seeing what adding this base class really buys us.

FWIW, I did expect this particular point to be that which we needed to iterate on to meet both our expectations.

Reply:
> Not quite. The NSBoundaryInterace includes an inviscid interface already, a boundary device should not inherit both. The NSBoundaryInterace already inherits the InviscidBoundaryInterace.

AFAIK there's no problem with inheriting both. They're just interfaces; no data.

> The thing that I do not like with (your) suggested approach is that it requires every boundary device to inherit the same N things. I prefer the approach of proving a base thing that is all N things, then inheriting that base thing in all the boundaries. As I mentioned, this centralizes the spot where we can come in and add operator support to existing boundaries.

Each boundary inherits according to which operators it's compatible with. It's not going to be the same for every BC. Heat BCs will be different, RTE Helmholtz BCs will be different, etc. I don't see why Euler and NS necessarily need to be lumped together. If the intent is just to avoid code duplication, the same can be achieved with helper functions without mandating that all fluid BCs must work for both Euler and NS. (Related question: does a slip boundary make sense in NS? This is the one I wasn't sure about originally.)

> FWIW, I did expect this particular point to be that which we needed to iterate on to meet both our expectations.

True, but the point of this discussion is to establish some shared convention for handling BCs for different operators and different physics, so in that sense it's relevant.

Reply:
> > FWIW, I did expect this particular point to be that which we needed to iterate on to meet both our expectations.
> 
> True, but the point of this discussion is to establish some shared convention for handling BCs for different operators and different physics, so in that sense it's relevant.

On this we agree completely.  I was only saying that this particular issue was one I expected to elicit a good deal of discussion before getting to a form we can both live with.